### DISCUSSION TITLE
Patch-based pipeline for 2D histology dataset

### DISCUSSION THREAD

**Discussion by Jarartur** [TorchIO-project/torchio#692] [Q&A] [answered]
Hi,
i don't have a lot of experience with torchIO and wanted to ask if the patch-based pipeline described [here](https://torchio.readthedocs.io/data/patch_based.html) will work with large 2D images.

Also, if it is possible:
To apply transformations and general preprocessing _(like affine aligment)_ I need to stick with torchIO transformations or can I use something like albumentations _(my current pipeline)_?

**Comment by romainVala** [comment, marked as answer]
Hello
yes the patch-based will work, I never test, but I think you just need to put a 1 in the third dimension
(you may get 3D patches, but with a 1 in the third dimension)

I do think you can easily use albumentations, but you should find similar transform in torchio ... are there some missing ?

**Reply by Jarartur** [reply]
I am currently using albumentations because it allows me to do the same random augmentations on two images. For example, when I apply random rotation, I need both source and target image to be rotated the same way _(like augmentations on an image and its mask basically)_.

**Reply by romainVala** [reply]
this is also the case with torchio
you have to construct a subject with 2 volume in it the input image (type torchio.INTENSITY ) and your target (type torchio.LABEL)
then for a random spatial transform (affine, crop ...) both will have the exact same transform, but for intensity (noise bias gamma ect ...) only the input image will be modify

and note also, that when creating patch, you will get of course the same patch for both

**Reply by romainVala** [reply]
but I just see, that you want to do a registration task, so it is not that easy than segmentation where your target is a mask of label ...

**Reply by Jarartur** [reply]
Thank you for providing details on how to do it in torchIO. Still the difference in intensity augmentations won't be ideal. 
I will certainly look more into it.

**Reply by fepegar** [reply]
I'm not sure what's the missing feature. Could you explain a bit more what you mean about the difference in intensity augmentations?

**Reply by Jarartur** [reply]
I was just referring to romainVala way of making my target something that won't accept intensity augmentations. If I make them 2 torchi.ScalarImages then I think It would be fine.

**Comment by fepegar** [comment]
Hi, @Jarartur. Why would you like to use TorchIO for 2D images? Understanding your use case will help us suggest solutions.

**Reply by Jarartur** [reply]
Yes, exactly that. I have a set of large 2D images from [here](https://anhir.grand-challenge.org/) and want to divide them in patches for learning but need to do inference on a whole image. As far as I understand it torchIO patchSampler and gridSampler do exactly that but by default for 3D. I am wondering if I can apply it for 2D.

**Reply by fepegar** [reply]
Sure, you can. You can use a patch size `(patch_w, patch_h, 1)`.

**Reply by fepegar** [reply]
On IPython (or Jupyter):

```python
import torchio as tio
!curl -o image.jpg https://thispersondoesnotexist.com/image
image = tio.ScalarImage('image.jpg')
image.as_pil().show()
```

![image](https://user-images.githubusercontent.com/12688084/138565777-64bb1131-c522-4cf9-a83f-5b0fc92dc577.jpg)

```python
from torchvision.utils import make_grid
import torchvision.transforms.functional as F
patch_size = 128, 128, 1
subject = tio.Subject(im=image)
sampler = tio.GridSampler(subject, patch_size)
tensors = [patch.im.data[..., 0] for patch in sampler]
grid = make_grid(tensors)
F.to_pil_image(grid).rotate(-90).show()
```

![patches](https://user-images.githubusercontent.com/12688084/138565980-5a8a2800-ca4b-4867-8afc-b645b7e2178a.png)

```python
from itertools import islice
sampler = tio.UniformSampler(patch_size)
patches = islice(sampler(subject), 64)
tensors = [patch.im.data[..., 0] for patch in patches]
grid = make_grid(tensors)
F.to_pil_image(grid).show()
```

![tmpxaesennk](https://user-images.githubusercontent.com/12688084/138565974-8f031973-3a0a-4775-a81f-5a659f787d76.png)

**Reply by fepegar** [reply]
I cheated a bit using `rotate` for visualization and I'm not sure why the patches are flipped. I'd need to think about it. But you see the idea.

**Reply by romainVala** [reply]
why can you not do the same with torchio, instead of torchvision ?

**Reply by Jarartur** [reply]
One problem in my approach is that I need to make patches from two images that need to be preprocessed based on each other before augmentations and unfortunately, I don't see how I would go about that.
From what I understand I would need to abandon tio.ScalarImage completely and at the same time I can't afford to load all images at once (so I need lazy loading).
Currently I am doing that by writing my custom Dataset from pytorch where I can load and preprocess two images before even going into Dataloader.

**Reply by fepegar** [reply]
What about using a transform that extracts patches?

```python
import torchio as tio


class SamplePatches:
    def __init__(self, patch_size, num_patches, image_name):
        self.patch_size = patch_size
        self.num_patches = num_patches
        self.image_name = image_name

    def __call__(self, subject):
        sampler = tio.UniformSampler(self.patch_size)
        patches = [p for p in sampler(subject, self.num_patches)]
        first = patches[0]
        for i, p in enumerate(patches[1:]):
            first.add_image(p[self.image_name], f'{self.image_name}_{i + 1}')
        return first


num_patches = 8
patch_size = 64

sheep = tio.datasets.Sheep()
sample_patches = SamplePatches(patch_size, num_patches, 't1')
sheep_patches = sample_patches(sheep)
sheep_patches.plot()
```

![Figure_1](https://user-images.githubusercontent.com/12688084/138603054-f73167ad-69bb-4ee5-8080-3fe6a6ed8763.png)

**Reply by Jarartur** [reply]
Yes I think that's one way to consider. I will try to create something like that but with consideration on the other [disscusion](https://github.com/fepegar/torchio/discussions/698).
I am more comfortable now with creating my custom transforms so I think this is the way I will go about doing that.

**Reply by romainVala** [reply]
May be it is not appropriate (not sure to follow properly here)

But here is an example, with torchio Queue, that perfectly mange the fact to get the exact same patch from 2 image 
```python

import torchio as tio
!curl -o image.jpg https://thispersondoesnotexist.com/image
image1 = tio.ScalarImage('image.jpg')
!curl -o image2.jpg https://thispersondoesnotexist.com/image
image2 = tio.ScalarImage('image2.jpg')

subject = tio.Subject(im1=image1, im2=image2)

suj_dataset = tio.SubjectsDataset([subject])
patch_size = 128, 128, 1
sampler = tio.GridSampler(subject, patch_size) #or  tio.UniformSampler(patch_size)

queue = tio.Queue(suj_dataset, sampler=sampler, max_length=64, samples_per_volume=64)

one_patch = queue[10]

#and you can then use the queue as your dataset in a standard torch DataLoader
```
